controller/backing_image_data_source_controller.go

```diff
@@ -0,0 +1,855 @@
+}
+
+func (c *BackingImageDataSourceController) isMonitoring(bidsName string) bool {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	_, ok := c.monitorMap[bidsName]
+	return ok
+}
+
+func (c *BackingImageDataSourceController) stopMonitoring(bidsName string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	log := c.logger.WithField("backingImageDataSource", bidsName)
+	stopCh, ok := c.monitorMap[bidsName]
+	if !ok {
+		log.Warn("No monitor goroutine for stopping")
+		return
+	}
+	log.Infof("Stopping monitoring")
+	close(stopCh)
+	delete(c.monitorMap, bidsName)
+	log.Infof("Stopped monitoring")
+
+	return
+}
+
+func (c *BackingImageDataSourceController) startMonitoring(bids *longhorn.BackingImageDataSource) {
+	log := getLoggerForBackingImageDataSource(c.logger, bids)
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if _, ok := c.monitorMap[bids.Name]; ok {
+		log.Error("BUG: Monitoring goroutine already exists")
+		return
+	}
+
+	pod, err := c.ds.GetPod(types.GetBackingImageDataSourcePodName(bids.Name))
+	if err != nil {
+		log.Errorf("Failed to get pod before launching the monitor: %v", err)
+		return
+	}
+	if pod == nil {
+		log.Errorf("Can not find pod before launching the monitor")
+		return
+	}
+
+	stopCh := make(chan struct{}, 1)
+	m := &BackingImageDataSourceMonitor{
+		Name:         bids.Name,
+		client:       engineapi.NewBackingImageDataSourceClient(pod.Status.PodIP),
+		stopCh:       stopCh,
+		controllerID: c.controllerID,
+		log:          log,
+		ds:           c.ds,
+		backoff:      c.backoff,
+	}
+	c.monitorMap[bids.Name] = stopCh
+
+	log.Infof("Start monitoring")
+
+	go wait.Until(m.sync, engineapi.BackingImageDataSourcePollInterval, stopCh)
+	go func() {
+		<-m.stopCh
+		c.stopMonitoring(bids.Name)
+	}()
+}
+
+func (m *BackingImageDataSourceMonitor) sync() {
+	var syncErr error
+	defer func() {
+		if syncErr != nil {
+			m.retryCount++
+			if m.retryCount == engineapi.MaxMonitorRetryCount {
+				close(m.stopCh)
+				logrus.Warnf("Stop monitoring since monitor %v sync reaches the max retry count %v", m.Name, engineapi.MaxMonitorRetryCount)
+				return
+			}
+		} else {
+			m.retryCount = 0
+		}
+
+	}()
+
+	bids, err := m.ds.GetBackingImageDataSource(m.Name)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			close(m.stopCh)
+			logrus.Warnf("Stop monitoring since backing image data source %v is not found", m.Name)
+			return
+		}
+		syncErr = errors.Wrapf(err, "failed to get backing image data source %v during monitor sync", m.Name)
+		logrus.Error(syncErr)
+		return
+	}
+	if bids.Status.OwnerID != m.controllerID {
+		close(m.stopCh)
+		logrus.Warnf("Stop monitoring since backing image data source %v owner %v is not the same as monitor current controller %v", m.Name, bids.Status.OwnerID, m.controllerID)
+		return
+	}
+
+	fileInfo, err := m.client.Get()
+	if err != nil {
+		syncErr = errors.Wrapf(err, "failed to get %v info from backing image data source server", m.Name)
+		logrus.Error(syncErr)
```

Comment:
NIT: any reason not to use m.log for these?

Reply:
No. I was just used to print the log via `logrus`. I will replace all `logrus` with `m.log`.